osutil/disks/disks.go

```diff
@@ -0,0 +1,293 @@
+	FindMatchingPartitionUUID(string) (string, error)
+
+	// MountPointIsFromDisk returns whether the specified mountpoint corresponds
+	// to a partition on the disk. Note that this only considers partitions
+	// and mountpoints found when the disk was identified with
+	// DiskFromMountPoint.
+	MountPointIsFromDisk(string, *Options) (bool, error)
+
+	// Dev returns the string "major:minor" for the disk device for
+	// identification, it should be unique but is not guaranteed to be unique.
+	Dev() string
+}
+
+type partition struct {
+	major    int
+	minor    int
+	label    string
+	partuuid string
+	path     string
+}
+
+type disk struct {
+	major      int
+	minor      int
+	partitions []*partition
+}
+
+func parseDeviceMajorMinor(s string) (int, int, error) {
+	errMsg := fmt.Errorf("invalid device number format: (expected <int>:<int>)")
+	devNums := strings.SplitN(s, ":", 2)
+	if len(devNums) != 2 {
+		return 0, 0, errMsg
+	}
+	maj, err := strconv.Atoi(devNums[0])
+	if err != nil {
+		return 0, 0, errMsg
+	}
+	min, err := strconv.Atoi(devNums[1])
+	if err != nil {
+		return 0, 0, errMsg
+	}
+	return maj, min, nil
+}
+
+var udevadmProperties = func(device string) ([]byte, error) {
+	cmd := exec.Command("udevadm", "info", "--query", "property", "--name", device)
+	return cmd.CombinedOutput()
+}
+
+func udevProperties(device string) (map[string]string, error) {
+	out, err := udevadmProperties(device)
+	if err != nil {
+		return nil, osutil.OutputErr(out, err)
+	}
+	r := bytes.NewBuffer(out)
+
+	return parseUdevProperties(r)
+}
+
+func parseUdevProperties(r io.Reader) (map[string]string, error) {
+	m := make(map[string]string)
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		strs := strings.SplitN(scanner.Text(), "=", 2)
+		if len(strs) != 2 {
+			// bad udev output?
+			continue
+		}
+		m[strs[0]] = strs[1]
+	}
+
+	return m, scanner.Err()
+}
+
+// DiskFromMountPoint finds a matching Disk for the specified mount point. It
+// does a best effort in identifying partitions for the disk, but may be racy
+// due to races inherent in the initramfs surrounding udev and sysfs.
+func DiskFromMountPoint(mountpoint string, opts *Options) (Disk, error) {
+	// call the unexported version that may be mocked by tests
+	return diskFromMountPoint(mountpoint, opts)
+}
+
+// diskFromMountPointImpl uses the mount table, sysfs and udev to try and
+// identify the disk and full set of partitions for the specified mount point.
+// since during the initramfs things may be racy around udev adding new devices
+// this is a best effort search and partially matching partitions which do not
+// have the full information are discarded from the returned disk.
+func diskFromMountPointImpl(mountpoint string, opts *Options) (*disk, error) {
+	// first get the mount entry for the mountpoint
+	mounts, err := osutil.LoadMountInfo()
+	if err != nil {
+		return nil, err
+	}
+	found := false
+	d := &disk{}
+	mountpointPart := partition{}
+	for _, mount := range mounts {
```

Comment:
I would recommend looking at the mount table in reverse. As otherwise you are matching on shadowed mounts.

Reply:
if there are extra complications maybe we should use "udevadm info --device-id-of-file" instead ? and work from its result?

Reply:
I guess since we already have the code to parse mount tables, why not use it? iterating in reverse is easy enough to do